File: linked list problems/Go/ADD_L.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
)

type Node struct {
	data int
	next *Node
}
func main() {
	
	// var head* Node = createLinkedList()
	// printLinkedList(head)
	// add_at_pos(head)
	// del_at_pos(head)
	// del_match(head)
	// reverse(head)
	// remove_duplicates_1(head)
	// print_mid(head)
	// create_loop(head)
	// nth_from_last(head)
	// merge_2_nodewise(head,head)

	var head1* Node = createLinkedList()
	printLinkedList(head1)
	var head2* Node = createLinkedList()
	printLinkedList(head2)

	merge_2_nodewise(head1,head2)

}

func del_at_pos(head* Node) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Please enter the position to delete");
	scanner.Scan()
	text := scanner.Text()
	pos, err := strconv.Atoi(text)
	temp := head
	if err != nil {
		return
	}
	curr_pos := 0

	for {

		if pos == 0 {
			head = head.next
			break
		}
		if temp.next !=nil {
			if curr_pos+1 == pos {
				curr_pos+=1
				temp.next = temp.next.next
				break
			} else {
				temp = temp.next
				curr_pos+=1
			}
		} else {
			break
		}
	}

	if pos<0 || curr_pos !=pos {
		fmt.Println("the index is out of bound")
	} else {
		fmt.Println("the data at the position is deleted")
	}

	printLinkedList(head)

}
func add_at_pos(head* Node) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Please enter the data to insert");
	scanner.Scan()
	text := scanner.Text()
	data, err := strconv.Atoi(text)
	fmt.Println("Please enter the position")
	scanner.Scan()
	pos_text := scanner.Text()
	pos, err := strconv.Atoi(pos_text)
	
	if err != nil {
		return
	}
	curr_pos := 0
	var new_node = Node{data,nil}
	temp := head

	for {
		if pos == 0 {
			new_node.next = head
			head = &new_node
			break
		}

		if temp != nil {

			if pos == curr_pos+1 {
				new_node.next = temp.next
				temp.next = &new_node
				curr_pos+=1
				break
			} else {
				temp = temp.next
				curr_pos+=1
			}
		} else {
			break
		}
	}

	if pos != curr_pos || pos<0 {
		fmt.Println("The index is out bound")
	}else {
		fmt.Println("The data has been inserted successfully")
	}
	printLinkedList(head)

}


func printLinkedList(head *Node){

	var temp *Node = head

	for {
		if temp==nil {
			break
		}
		fmt.Print(temp.data," ")
		temp = temp.next
	}
	fmt.Println()

}

func del_match(head* Node) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Please enter the data to delete");
	scanner.Scan()
	text := scanner.Text()
	data, err := strconv.Atoi(text)

	if err != nil {
		return
	}

	curr_pos := 0
	temp := head

	if head.data == data {
		head = head.next
		fmt.Println("The data is found at ",curr_pos," and deleted")
		printLinkedList(head)
		return
	}

	for {
		if temp.next!=nil {

			if temp.next.data == data {
				curr_pos+=1;
				temp.next = temp.next.next
				fmt.Println("The data is found at ",curr_pos, " and deleted")
				printLinkedList(head)
				return
			} else {
				temp = temp.next
				curr_pos+=1
			}
		} else {
			break
		}
	}

	fmt.Println("The data is not found in the list")
}

func reverse(head* Node) {

	var next* Node = nil
	var prev* Node = nil

	curr:= head

	for {
		if curr.next!=nil {
			next = curr.next
			curr.next = prev
			prev = curr
			curr = next
		} else {
			curr.next = prev
			head = curr
			break
		}
	}

	fmt.Println("The reversed list is ")
	printLinkedList(head)
}


func remove_duplicates_1(head* Node) {

	temp:= head
	curr:= head

	for {
		if curr.next!=nil {
			for {
				fmt.Println("curr2",curr,temp)
				if temp.next!= nil{
					if curr.data == temp.next.data {
						temp.next = temp.next.next
					} else {
						temp = temp.next
					}
				} else {
					break
				}
			}
			if curr.next!=nil {
				curr = curr.next
				temp = curr
			}
			
		} else {
			break
		}
	}

	fmt.Println("The final list is ")
	printLinkedList(head)
}


func print_mid(head* Node) {
	slow:=head
	fast:=head
	for {
		if fast.next!=nil {
				
				fast = fast.next.next
				
				if fast==nil {
					fmt.Println("The middle node is ",slow.data)
					break
				}
				slow = slow.next

		} else {
				fmt.Println("The middle node is ",slow.data)
				break
			}
	}
}


func create_loop(head* Node) {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Enter the position to create a loop")
	scanner.Scan()
	text := scanner.Text()

	pos, err := strconv.Atoi(text)

	if err != nil {
		return
	}

	temp := head
	var loop* Node = nil
	curr_pos := 0

	for {

		if curr_pos==pos {
			loop = temp
		}

		if temp.next!=nil {
			temp = temp.next
			curr_pos+=1
		} else {
			temp.next = loop
			break
		}

	}

	if loop!=nil {
		fmt.Println("The loop is created and the loop node is ", loop.data)
		find_cycle(head)
		// printLinkedList(head)
	} else {
		fmt.Println("loop is not created")
		printLinkedList(head)

	}
}


func find_cycle(head* Node) {

	slow:=head
	fast:=head.next


	for {
		if fast != nil || fast.next!=nil {
			fast = fast.next.next
			slow = slow.next
			
			if slow == fast {
				fmt.Println(slow,fast)
				fmt.Println("the loop is found")
				break
			}
		} else  {
			fmt.Println("loop does not exist")
			break
		}
	}
}


func nth_from_last(head* Node){
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Enter the position to find the node from last")
	scanner.Scan()
	text := scanner.Text()

	pos, err := strconv.Atoi(text)
	curr_pos:=0

	if err!=nil {
		return
	}

	slow := head
	fast:=head
	for {
		if fast!=nil{

			for {
				if curr_pos > pos {
					break
				} else {
					fast = fast.next
					curr_pos+=1
				}
			}
				
			slow = slow.next
			fast = fast.next

		} else if slow!=head {
			fmt.Println("The nth from last is ",slow.data)
			break
		} else {
			fmt.Println("Out of range")
			break
		}
	}
}

func merge_2_nodewise(head1* Node,head2* Node) {

	curr1:=head1
	curr2:=head2
	var next1* Node = nil
	var next2* Node = nil

	for {

		if curr1!=nil && curr2!=nil {
			
			next1= curr1.next
			next2= curr2.next

			curr1.next =curr2
			curr2.next = next1

			curr1 = next1
			if next1!=nil{
				curr2 = next2
			} else {
				curr2.next = next2 
			}

		} else{
			break
		}
			
	}

	fmt.Println("Merged list is ")
	printLinkedList(head1)

}


func createLinkedList() *Node {

	var head* Node = nil
	fmt.Println("Please continue entering the data to the linkedlist and type done to stop it\n");

	scanner := bufio.NewScanner(os.Stdin)

	for {
		scanner.Scan()
		text := scanner.Text()

		if text == "done" {
			break
		}
		data, err := strconv.Atoi(text)
		
		if err == nil {
        	var new_node = Node{data,nil}

        	if head == nil {
        		head = &new_node
        	} else {
        		temp := head
        		for{
        			if temp.next==nil {
	        			temp.next = &new_node
	        			break
        			}
        			temp = temp.next
        			
        		}
        	}
		fmt.Println("New node",data," is added to the linked list and type done to stop it")

    	}

	}

	return head
}

File: linked list problems/C++/ADD_L.cpp
#include <iostream>

using namespace std;

struct Node
{
	int data;
	struct Node *next = NULL;
};

void printLinkedList(struct Node* head )
{
	struct Node* itr = head;
			
	while(itr!=NULL)
	{
		cout<<itr->data<<" ";
		itr = itr->next;
	}
	cout<<endl;
}

void reverse(struct Node* head)
{
	struct Node* curr = head;
	struct Node* prev = NULL;
	struct Node* next = NULL;

	while(curr->next!=NULL)
	{
		next = curr->next;
		curr->next = prev;
		prev = curr;
		curr = next;
	}
	curr->next = prev;
	head = curr;
	cout<<"The reversed list is "<<endl;
	printLinkedList(head);
}

void add_at_pos(struct Node *head)
	{
		cout<<"Please enter the data to insert"<<endl;
		int data;
		cin>>data;
		cout<<"please enter the position"<<endl;
		int pos;
		cin>>pos;

		struct Node *temp = head;
		int curr_pos = 0;
		struct Node *new_node = (struct Node*) malloc (sizeof(struct Node));
		new_node->data = data;
		new_node->next = NULL;

		while(temp!=NULL)
		{
			if (pos==0)
			{
				new_node->next = head;
				head = new_node;
				break;
			}

			if(curr_pos+1==pos)
			{
				new_node->next = temp->next;
				temp->next = new_node;
				curr_pos+=1;
				break;
			}
			else
			{
				temp = temp->next;
				curr_pos += 1;
			}

		}

		if ( (pos<0) || (curr_pos!=pos) )
			cout<<"The index is out bound"<<endl;
		else
			cout<<"The data has been inserted successfully"<<endl;
	
		printLinkedList(head);

	}


void del_at_pos(struct Node *head)
{

	cout<<"please enter the position to delete an item"<<endl;
	int pos;
	cin>>pos;
	int curr_pos = 0;

	struct Node *temp = head;

	while(temp->next!=NULL)
	{
		if (pos==0)
		{
			head = head->next;
			break;
		}
		if (curr_pos+1==pos)
		{
			temp->next = temp->next->next;
			curr_pos+= 1;
			break;
		}
		else
		{
			temp= temp->next;
			curr_pos+= 1;
		}
	}

	if ((pos<0) || (curr_pos!=pos) )
		cout<<"the index is out of bound"<<endl;
	else
		cout<<"The data at the position is deleted"<<endl;

	printLinkedList(head);

}

void del_match(struct Node *head)
{
	cout<<"please enter the data to delete"<<endl;
	int data;
	cin>>data;
	int pos=0;
	if(head->data == data)
	{
		head = head->next;
		cout<<"The data has been found at pos "<<pos<<" and deleted"<<endl;
		printLinkedList(head);
		return;
	}
	struct Node *temp = head;

	while(temp->next!=NULL)
	{

		if (temp->next->data == data)
		{	
			pos+=1;
			temp->next = temp->next->next;
			cout<<"The data has been found at pos "<<pos<<" and deleted"<<endl;
			printLinkedList(head);
			return;
		}
		else
		{
			temp = temp->next;
			pos+= 1;

		}		
	}
	
	cout<<"data is not found in the list"<<endl;

}

void remove_duplicates(struct Node *head)
{
	struct Node *curr = head;
	struct Node *temp = head;

	while (curr->next!=NULL)
	{
		while(temp->next!=NULL)
		{
			if (temp->next->data == curr->data)
			{
				temp->next = temp->next->next;
			}
			else
			{
				temp = temp->next;
			}
		}

		if (curr->next!=NULL)
		{
			curr = curr->next;
			temp = curr;
		}
	}
	cout<<"The final list is "<<endl;
	printLinkedList(head);
}

void print_mid(struct Node *head) 
{
	struct Node *slow = head;
	struct Node *fast = head;

	while (fast->next!=NULL)
	{
		slow = slow->next;
		fast = fast->next->next;
		if (fast==NULL)
			break;
	}

	cout<<"the middle element in the list is " <<slow->data<<endl;
}


struct Node* createLinkedList()
{
	struct Node *head = NULL;
	bool done = false;

	cout<<"Please continue entering the data to the linkedlist and type done to stop it\n";
	
	while(!done)
	{
		string input;	
		cin >> input;

		if (input=="done")
			break;

		int data = stoi(input);

		struct Node *temp = (struct Node*) malloc (sizeof(struct Node));
		temp->data = data;
		temp->next = NULL;

		if(head==NULL)
		{
			head= temp;
		}
		else
		{
			struct Node* itr = head;
			
			while(itr->next!=NULL)
			{
				itr = itr->next;
			}

			itr->next = temp;
		}

		cout<<"New node "<<data<<" is added to the linked list and type done to stop it"<<endl;

	}

	// printLinkedList(head);
	// add_at_pos(head);
	// del_at_pos(head);
	// del_match(head);
	// reverse(head);
	// remove_duplicates(head);
	return head;
	// print_mid(head);

}

void find_loop(struct Node *head)
{
	struct Node *slow = head;
	struct Node *fast = head->next;

	while(fast->next!=NULL)
	{
		slow = slow->next;
		fast = fast->next->next;

		if (fast==slow)
		{
			cout<<"The loop is found at "<<fast->data<<endl;
			break;
		}
	}
}
void create_loop(struct Node *head)
{
	struct Node *loop_node = NULL;
	struct Node *temp = head;
	cout<<"Please enter the index where you wanna create a loop"<<endl;

	int index;
	int curr_pos =0;
	cin>>index;

	while(temp->next!=NULL)
	{
		if (curr_pos==index)
		{
			loop_node = temp;
		}
		temp = temp->next;
		curr_pos+=1;
	}

	if (curr_pos==index)
		loop_node = temp;

	temp->next = loop_node;
	cout<<"The loop node is "<<loop_node->data<<endl;
	find_loop(head);
}

int main()
{
	struct Node *head = createLinkedList();
	printLinkedList(head);
	create_loop(head);

	return 0;
}